gobblin-runtime/src/main/java/gobblin/runtime/api/JobExecutionState.java

```diff
@@ -51,18 +60,51 @@ public JobExecutionState(JobSpec jobSpec, JobExecution jobExecution,
   }
 
   public Map<String, Object> getExecutionMetadata() {
-    return Collections.unmodifiableMap(this.executionMetadata);
+    this.changeLock.lock();
+    try {
+      return Collections.unmodifiableMap(this.executionMetadata);
+    }
+    finally {
+      this.changeLock.unlock();
+    }
+  }
+
+  @Override public String toString() {
+    this.changeLock.lock();
+    try {
+      return Objects.toStringHelper(JobExecutionState.class.getSimpleName())
+              .add("jobExecution", this.jobExecution)
+              .add("runningState", this.runningState)
+              .add("stage", this.stage)
+              .toString();
+    }
+    finally {
+      this.changeLock.unlock();
+    }
+  }
+
+  @Override public JobState.RunningState getRunningState() {
+    this.changeLock.lock();
+    try {
+      return this.runningState;
```

Comment:
Yes, both read and write paths need to be protected for consistency. We can try to avoid using lock with "transient" but in some perf experiments I did once, I didn't find it much faster.

Better safe than sorry :-)